File: cvdl-rs/src/layout.rs
use std::collections::HashMap;
use std::fmt::Display;

use serde::{Deserialize, Serialize};

use crate::{
    alignment::Alignment,
    any_layout::ElementBox,
    basic_layout::BasicLayout,
    container::Container,
    element::Element,
    font::{Font, FontDict},
    margin::Margin,
    point::Point,
    resume_data::ItemContent,
    spatial_box::SpatialBox,
    width::Width,
};

#[derive(Serialize, Deserialize, Clone, Debug)]
pub enum SectionLayout {
    Stack(Container),
    FrozenRow(Container),
    FlexRow(Container),
    Text(Element),
    Ref(Element),
}

// todo: use this
impl From<SectionLayout> for BasicLayout {
    fn from(value: SectionLayout) -> Self {
        match value {
            SectionLayout::Stack(_container) => todo!(),
            SectionLayout::FrozenRow(_) => unreachable!("FrozenRow should be converted to FlexRow"),
            SectionLayout::FlexRow(_container) => todo!(),
            SectionLayout::Text(_element) => todo!(),
            SectionLayout::Ref(_) => unreachable!("Ref should be converted to Text"),
        }
    }
}

impl Display for SectionLayout {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            SectionLayout::Stack(container) => write!(f, "{}", container),
            SectionLayout::FrozenRow(container) => write!(f, "{}", container),
            SectionLayout::FlexRow(container) => write!(f, "{}", container),
            SectionLayout::Text(element) => write!(f, "{}", element),
            SectionLayout::Ref(element) => write!(f, "{}", element),
        }
    }
}

impl SectionLayout {
    pub fn new_stack(container: Container) -> SectionLayout {
        log::debug!("Creating new stack: {}", container.uid);
        SectionLayout::Stack(container)
    }

    pub fn new_frozen_row(container: Container) -> SectionLayout {
        log::debug!("Creating new frozen row: {}", container.uid);
        SectionLayout::FrozenRow(container)
    }

    pub fn new_flex_row(container: Container) -> SectionLayout {
        log::debug!("Creating new flex row: {}", container.uid);
        SectionLayout::FlexRow(container)
    }

    pub fn new_text(element: Element) -> SectionLayout {
        log::debug!("Creating new text element: {}", element.uid);
        SectionLayout::Text(element)
    }

    pub fn new_ref(element: Element) -> SectionLayout {
        log::debug!("Creating new ref element: {}", element.uid);
        SectionLayout::Ref(element)
    }
}

impl SectionLayout {
    #[allow(dead_code)]
    pub fn type_(&self) -> String {
        match self {
            SectionLayout::Stack(_) => "stack".to_string(),
            SectionLayout::FrozenRow(_) => "frozen_row".to_string(),
            SectionLayout::FlexRow(_) => "flex_row".to_string(),
            SectionLayout::Text(_) => "text".to_string(),
            SectionLayout::Ref(_) => "ref".to_string(),
        }
    }

    pub fn width(&self) -> Width {
        match self {
            SectionLayout::Stack(container)
            | SectionLayout::FrozenRow(container)
            | SectionLayout::FlexRow(container) => container.width,
            SectionLayout::Text(element) | SectionLayout::Ref(element) => element.width,
        }
    }
    #[allow(dead_code)]
    pub fn margin(&self) -> Margin {
        match self {
            SectionLayout::Stack(container)
            | SectionLayout::FrozenRow(container)
            | SectionLayout::FlexRow(container) => container.margin,
            SectionLayout::Text(element) | SectionLayout::Ref(element) => element.margin,
        }
    }
    #[allow(dead_code)]
    pub fn alignment(&self) -> Alignment {
        match self {
            SectionLayout::Stack(container)
            | SectionLayout::FrozenRow(container)
            | SectionLayout::FlexRow(container) => container.alignment,
            SectionLayout::Text(element) | SectionLayout::Ref(element) => element.alignment,
        }
    }

    pub fn fonts(&self) -> Vec<Font> {
        match self {
            SectionLayout::Stack(container)
            | SectionLayout::FrozenRow(container)
            | SectionLayout::FlexRow(container) => container.fonts(),
            SectionLayout::Text(element) | SectionLayout::Ref(element) => {
                vec![element.font.clone()]
            }
        }
    }
    #[allow(dead_code)]
    pub fn with_margin(&self, margin: Margin) -> SectionLayout {
        match self {
            SectionLayout::Stack(container) => {
                SectionLayout::new_stack(container.with_margin(margin))
            }
            SectionLayout::FrozenRow(container) => {
                SectionLayout::new_frozen_row(container.with_margin(margin))
            }
            SectionLayout::FlexRow(container) => {
                SectionLayout::new_flex_row(container.with_margin(margin))
            }
            SectionLayout::Text(element) => SectionLayout::new_text(element.with_margin(margin)),
            SectionLayout::Ref(element) => SectionLayout::new_ref(element.with_margin(margin)),
        }
    }
    #[allow(dead_code)]
    pub fn with_alignment(&self, alignment: Alignment) -> SectionLayout {
        match self {
            SectionLayout::Stack(container) => {
                SectionLayout::new_stack(container.with_alignment(alignment))
            }
            SectionLayout::FrozenRow(container) => {
                SectionLayout::new_frozen_row(container.with_alignment(alignment))
            }
            SectionLayout::FlexRow(container) => {
                SectionLayout::new_flex_row(container.with_alignment(alignment))
            }
            SectionLayout::Text(element) => {
                SectionLayout::new_text(element.with_alignment(alignment))
            }
            SectionLayout::Ref(element) => {
                SectionLayout::new_ref(element.with_alignment(alignment))
            }
        }
    }

    pub fn is_instantiated(&self) -> bool {
        log::debug!("Checking if {} is instantiated...", self);
        match self {
            SectionLayout::Stack(c) | SectionLayout::FrozenRow(c) | SectionLayout::FlexRow(c) => {
                c.elements.iter().all(|e| e.is_instantiated())
            }
            SectionLayout::Text(_) => true,
            SectionLayout::Ref(_) => false,
        }
    }

    pub fn instantiate(&self, section: &HashMap<String, ItemContent>) -> SectionLayout {
        match self {
            SectionLayout::Stack(c) => SectionLayout::new_stack(c.instantiate(section)),
            SectionLayout::FrozenRow(c) => SectionLayout::new_frozen_row(c.instantiate(section)),
            SectionLayout::FlexRow(c) => SectionLayout::new_flex_row(c.instantiate(section)),
            SectionLayout::Text(e) => SectionLayout::new_text(e.clone()),
            SectionLayout::Ref(e) => SectionLayout::instantiate_ref_element(e.clone(), section),
        }
    }

    pub fn instantiate_ref_element(
        element: Element,
        section: &HashMap<String, ItemContent>,
    ) -> SectionLayout {
        if let Some(text) = section.get(&element.item) {
            let mut element = element.with_item(text.to_string());

            if let ItemContent::Url { url, text: _ } = text {
                element = element.with_url(url.clone())
            }

            SectionLayout::Text(element)
        } else {
            SectionLayout::Stack(Container::empty_container())
        }
    }

    pub fn bound_width(&self, width: f32) -> SectionLayout {
        let bound = match self.width() {
            Width::Absolute(w) => f32::min(w, width),
            Width::Percentage(_) => unreachable!(
                "SectionLayout::bound_width: Cannot bounded width for non-unitized widths!"
            ),
            Width::Fill => width,
        };

        match self {
            SectionLayout::Stack(c) => SectionLayout::new_stack(c.bound_width(bound)),
            SectionLayout::FrozenRow(c) => SectionLayout::new_frozen_row(c.bound_width(bound)),
            SectionLayout::FlexRow(c) => SectionLayout::new_flex_row(c.bound_width(bound)),
            SectionLayout::Text(e) => SectionLayout::new_text(e.bound_width(bound)),
            SectionLayout::Ref(_) => {
                unreachable!("Cannot propagate widths of uninstantiated layout")
            }
        }
    }

    pub fn scale_width(&self, document_width: f32) -> SectionLayout {
        match self {
            SectionLayout::Stack(c) => SectionLayout::new_stack(c.scale_width(document_width)),
            SectionLayout::FrozenRow(c) => {
                SectionLayout::new_frozen_row(c.scale_width(document_width))
            }
            SectionLayout::FlexRow(c) => SectionLayout::new_flex_row(c.scale_width(document_width)),
            SectionLayout::Text(e) => SectionLayout::new_text(e.scale_width(document_width)),
            SectionLayout::Ref(_) => unreachable!("Cannot scale width of uninstantiated layout"),
        }
    }

    pub fn normalize(&self, width: f32, font_dict: &FontDict) -> SectionLayout {
        log::debug!(
            "Normalizing document, checking if {} is instantiated...",
            self
        );

        if !self.is_instantiated() {
            log::error!("Document is not instantiated {}", self);
            panic!("Cannot normalize uninstantiated layout");
        };

        log::debug!("Document is instantiated. Scaling widths...");

        let scaled_layout = self.scale_width(width);

        log::debug!("Widths are scaled. Bounding widths...");

        let bounded_layout = scaled_layout.bound_width(width);

        log::debug!("Widths are bounded. Filling fonts...");

        let font_filled_layout = bounded_layout.fill_fonts(font_dict);

        log::debug!("Fonts filled. Breaking lines...");

        let broken_layout = font_filled_layout.break_lines(font_dict);

        log::debug!("Lines broken.");

        broken_layout
    }

    pub fn fill_fonts(&self, font_dict: &FontDict) -> SectionLayout {
        match self {
            SectionLayout::Stack(c) => SectionLayout::new_stack(c.fill_fonts(font_dict)),
            SectionLayout::FrozenRow(c) => SectionLayout::new_frozen_row(c.fill_fonts(font_dict)),
            SectionLayout::FlexRow(c) => SectionLayout::new_flex_row(c.fill_fonts(font_dict)),
            SectionLayout::Text(e) => SectionLayout::new_text(e.fill_fonts(font_dict)),
            SectionLayout::Ref(_) => unreachable!("Cannot fill fonts of uninstantiated layout"),
        }
    }

    pub fn break_lines(&self, font_dict: &FontDict) -> SectionLayout {
        match self {
            SectionLayout::Stack(c) => {
                let new_stack = SectionLayout::new_stack(
                    c.with_elements(
                        c.elements
                            .iter()
                            .map(|e| e.break_lines(font_dict))
                            .collect(),
                    ),
                );
                new_stack
            }
            SectionLayout::FrozenRow(c) => {
                let total_width = c
                    .elements
                    .iter()
                    .map(|e| e.width().get_fixed_unchecked())
                    .sum::<f32>();
                if total_width > self.width().get_fixed_unchecked() {
                    panic!(
                        "Cannot break lines of frozen row with width {:?} and total width {}",
                        self.width(),
                        total_width
                    );
                } else {
                    SectionLayout::new_flex_row(Container {
                        uid: c.uid,
                        elements: c
                            .elements
                            .iter()
                            .map(|e| e.break_lines(font_dict))
                            .collect(),
                        margin: c.margin,
                        alignment: c.alignment,
                        width: c.width,
                    })
                }
            }
            SectionLayout::FlexRow(c) => {
                let lines: Vec<Container> = c.break_lines(font_dict);
                SectionLayout::new_stack(
                    c.with_elements(lines.into_iter().map(SectionLayout::FlexRow).collect()),
                )
            }
            SectionLayout::Text(e) => {
                let lines: Vec<Element> = e.break_lines(font_dict);
                SectionLayout::new_stack(
                    Container::empty_container()
                        .with_elements(lines.into_iter().map(SectionLayout::new_text).collect())
                        .with_alignment(e.alignment)
                        .with_margin(e.margin)
                        .with_width(e.width),
                )
            }
            SectionLayout::Ref(_) => unreachable!("Cannot break lines of uninstantiated layout"),
        }
    }
}

impl SectionLayout {
    pub fn compute_boxes(&self, font_dict: &FontDict) -> ElementBox {
        let mut textbox_positions: Vec<(SpatialBox, Element)> = Vec::new();
        let top_left: Point = Point::new(0.0, 0.0);
        let depth = self.compute_textbox_positions(&mut textbox_positions, top_left, font_dict);

        let bounding_box = SpatialBox::new(
            Point::new(0.0, 0.0),
            Point::new(self.width().get_fixed_unchecked(), depth),
        );
        ElementBox::new(bounding_box, textbox_positions)
    }

    fn compute_textbox_positions(
        &self,
        textbox_positions: &mut Vec<(SpatialBox, Element)>,
        top_left: Point,
        font_dict: &FontDict,
    ) -> f32 {
        match self {
            SectionLayout::Stack(c) => {
                let mut top_left = top_left;
                let mut depth = top_left.y;
                for element in c.elements.iter() {
                    depth =
                        element.compute_textbox_positions(textbox_positions, top_left, font_dict);
                    top_left = top_left.move_y_to(depth);
                }
                depth
            }
            SectionLayout::FlexRow(c) => {
                let (top_left, per_elem_space) = match c.alignment {
                    Alignment::Left => (top_left, 0.0),
                    Alignment::Center => (
                        top_left
                            .move_x_by((c.width.get_fixed_unchecked() - c.elements_width()) / 2.0),
                        0.0,
                    ),
                    Alignment::Right => (
                        top_left.move_x_by(c.width.get_fixed_unchecked() - c.elements_width()),
                        0.0,
                    ),
                    Alignment::Justified => (
                        top_left,
                        (c.width.get_fixed_unchecked() - c.elements_width())
                            / (c.elements.len() - 1) as f32,
                    ),
                };

                let mut top_left = top_left;
                let mut depth = top_left.y;

                for element in c.elements.iter() {
                    depth =
                        element.compute_textbox_positions(textbox_positions, top_left, font_dict);
                    top_left =
                        top_left.move_x_by(element.width().get_fixed_unchecked() + per_elem_space);
                }
                depth
            }
            SectionLayout::FrozenRow(_) => {
                unreachable!("Cannot compute textbox positions of frozen row: {:?}", self)
            }
            SectionLayout::Text(e) => {
                let width = e.text_width.get_fixed_unchecked();
                let height = e.font.get_height(font_dict);
                let textbox =
                    SpatialBox::new(top_left, top_left.move_x_by(width).move_y_by(height));
                textbox_positions.push((textbox, e.clone()));

                top_left.y + height
            }
            SectionLayout::Ref(_) => {
                todo!("Should not be able to compute textbox positions of uninstantiated layout")
            }
        }
    }
}

File: cvdl-ts/src/Layout.ts
import * as Width from "./Width";
import * as Font from "./Font";
import { Box } from "./Box";
import { ElementPath, FontDict } from "./AnyLayout";
import { Point } from "./Point";
import * as Stack from "./Stack";
import * as Row from "./Row";
import * as Elem from "./Elem";
import { Field } from "./DataSchema";
import { with_ } from "./Utils";
import * as Alignment from "./Alignment";
import * as Margin from "./Margin";
import * as Resume from "./Resume";

export type Container = Stack.t | Row.t;
export type t = Stack.t | Row.t | Elem.t;
type Layout = t;

// A rendered layout is a layout with a bounding box and where all the elements are also rendered layouts
export type RenderedStack = Stack.t & {
	bounding_box: Box;
	elements: RenderedLayout[];
};
export type RenderedRow = Row.t & {
	bounding_box: Box;
	elements: RenderedLayout[];
};

export type Binding = {
	binding: string;
};

export type PreBindingElem = Elem.t & {
	width: Width.t | Binding;
	font: Font.t | Binding;
	margin: Margin.t | Binding;
	alignment: Alignment.t | Binding;
	background_color: Color | Binding;
};

export type PreBindingContainer = Container & {
	width: Width.t | Binding;
	margin: Margin.t | Binding;
	alignment: Alignment.t | Binding;
};

export type PreBindingLayout = PreBindingElem | PreBindingContainer;
export type BindedLayout = Layout;

export type RenderedElem = Elem.t & { bounding_box: Box };

export type RenderedLayout = (RenderedStack | RenderedRow | RenderedElem) & {
	path?: ElementPath;
};

export function default_(tag: string) {
	switch (tag) {
		case "Stack":
			return Stack.default_();
		case "FlexRow":
			return Row.default_();
		case "FrozenRow":
			return with_(Row.default_(), { is_frozen: true });
		case "Text":
			return Elem.default_();
		case "Ref":
			return Elem.asRef(Elem.default_());
	}
}

export function empty(): Layout {
	return default_("Stack");
}

export function type_(l: Layout): string {
	return l.tag;
}

export function tag_(l: Layout): string {
	switch (l.tag) {
		case "Stack":
			return "Stack";
		case "Row": {
			const row = l as Row.t;
			return row.is_frozen ? "FrozenRow" : "FlexRow";
		}
		case "Elem": {
			const elem = l as Elem.t;
			return elem.is_ref ? "Ref" : "Text";
		}
	}
}

export function isContainer(l: Layout): boolean {
	return l.tag === "Stack" || l.tag === "Row";
}

export function isFill(l: Layout): boolean {
	switch (l.tag) {
		case "Stack":
			return (
				l.is_fill &&
				(l as Stack.t).elements
					.map((e) => isFill(e))
					.reduce((a, b) => a && b, true)
			);
		case "Row":
			return (
				l.is_fill &&
				(l as Row.t).elements
					.map((e) => isFill(e))
					.reduce((a, b) => a && b, true)
			);
		case "Elem":
			return l.is_fill;
	}
}

export function isRef(l: Layout): boolean {
	return l.tag === "Elem" && l.is_ref;
}

export function fonts(l: Layout): Font.t[] {
	switch (l.tag) {
		case "Stack":
			return (l as Stack.t).elements
				.map((e) => fonts(e))
				.reduce((a, b) => a.concat(b), []);
		case "Row":
			return (l as Row.t).elements
				.map((e) => fonts(e))
				.reduce((a, b) => a.concat(b), []);
		case "Elem":
			return [(l as Elem.t).font];
	}
}

export function totalElementsWidth(l: Layout): number {
	switch (l.tag) {
		case "Stack":
			return (l as Stack.t).elements
				.map((e) => totalElementsWidth(e))
				.reduce((a, b) => a + b, 0.0);
		case "Row":
			return (l as Row.t).elements
				.map((e) => totalElementsWidth(e))
				.reduce((a, b) => a + b, 0.0);
		case "Elem":
			return Width.get_fixed_unchecked((l as Elem.t).width);
	}
}

export function isInstantiated(l: Layout): boolean {
	if (isContainer(l)) {
		return (l as Container).elements
			.map((e) => isInstantiated(e))
			.reduce((a, b) => a && b, true);
	} else {
		return !isRef(l);
	}
}

export function instantiate(
	l: PreBindingLayout,
	section: Resume.Item,
	fields: Field[],
	bindings: Map<string, unknown>,
): Layout {
	switch (l.tag) {
		case "Stack":
		case "Row":
			return with_(l, {
				elements: (l as Container).elements.map((e) =>
					instantiate(e, section, fields, bindings),
				),
			});
		case "Elem":
			return Elem.instantiate(l, section, fields, bindings);
	}
}

export function boundWidth(l: Layout, width: number): Layout {
	const bound =
		l.width.tag === "Absolute"
			? Math.min(l.width.value, width)
			: l.width.tag === "Fill"
				? width
				: null;

	if (bound === null) {
		throw new Error("Cannot bound width of non-unitized widths!");
	}

	if (l.tag === "Elem" && l.is_ref) {
		throw new Error("Cannot propagate widths of uninstantiated layout");
	}

	switch (l.tag) {
		case "Stack":
			return Stack.boundWidth(
				l as Stack.t,
				bound - l.margin.left - l.margin.right,
			);
		case "Row":
			return Row.boundWidth(l as Row.t, bound - l.margin.left - l.margin.right);
		case "Elem":
			return Elem.boundWidth(
				l as Elem.t,
				bound - l.margin.left - l.margin.right,
			);
	}
}

export function scaleWidth(l: Layout, document_width: number): Layout {
	if (isRef(l)) {
		throw new Error("Cannot scale width of uninstantiated layout");
	}
	switch (l.tag) {
		case "Stack":
			return Stack.scaleWidth(l as Stack.t, document_width);
		case "Row":
			return Row.scaleWidth(l as Row.t, document_width);
		case "Elem":
			return Elem.scaleWidth(l as Elem.t, document_width);
	}
}

export function normalize(
	l: Layout,
	width: number,
	font_dict: FontDict,
): Layout {
	console.debug(
		`Normalizing document, checking if document is instantiated...`,
	);
	if (!isInstantiated(l)) {
		throw Error(`Cannot normalize uninstantiated layout ${JSON.stringify(l)}`);
	}

	console.debug("Document is instantiated. Scaling widths...");

	const scaled_layout = scaleWidth(l, width);

	console.debug("Widths are scaled. Bounding widths...");

	const bounded_layout = boundWidth(scaled_layout, width);

	console.debug("Widths are bounded. Filling fonts...");

	const font_filled_layout = fillFonts(bounded_layout, font_dict);

	console.debug("Fonts filled.");

	return font_filled_layout;
}

export function fillFonts(l: Layout, font_dict: FontDict): Layout {
	if (isRef(l)) {
		throw new Error("Cannot fill fonts of uninstantiated layout");
	}

	switch (l.tag) {
		case "Stack":
		case "Row": {
			const filledFonts = (l as Container).elements.map((e) =>
				fillFonts(e, font_dict),
			);
			l = { ...l, elements: filledFonts };
			if (isFill(l)) {
				const total_width = totalElementsWidth(l);
				if (total_width <= Width.get_fixed_unchecked(l.width)) {
					// throw `Cannot fill fonts of row with width ${JSON.stringify(this.width())} and total width ${total_width}`
					return with_(l, { width: Width.absolute(total_width) });
				}
			}
			return l;
		}
		case "Elem":
			return Elem.fillFonts(l as Elem.t, font_dict);
	}
}

export function computeBoxes(l: Layout, font_dict: FontDict): RenderedLayout {
	const top_left: Point = new Point(0.0, 0.0);
	return computeTextboxPositions(l, top_left, font_dict).renderedLayout;
}

export function computeTextboxPositions(
	l: Layout,
	top_left: Point,
	font_dict: FontDict,
): { depth: number; renderedLayout: RenderedLayout } {
	let depth = top_left.y;
	switch (l.tag) {
		case "Stack": {
			const stack = l as Stack.t;
			top_left = top_left
				.move_y_by(stack.margin.top)
				.move_x_by(stack.margin.left);
			const originalTopLeft = top_left;
			const renderedElements = [];
			for (const element of stack.elements) {
				const result = computeTextboxPositions(element, top_left, font_dict);
				depth = result.depth;

				top_left = top_left.move_y_to(depth);
				renderedElements.push(result.renderedLayout);
			}
			depth += stack.margin.bottom;
			return {
				depth,
				renderedLayout: {
					...l,
					bounding_box: new Box(
						originalTopLeft,
						top_left
							.move_x_by(Width.get_fixed_unchecked(stack.width))
							.move_y_by(stack.margin.bottom),
					),
					elements: renderedElements,
				},
			};
		}
		case "Row": {
			const row = l as Row.t;
			top_left = top_left.move_y_by(row.margin.top).move_x_by(row.margin.left);
			const originalTopLeft = top_left;
			let per_elem_space: number = 0.0;
			switch (row.alignment) {
				case "Center":
					top_left = top_left.move_x_by(
						(Width.get_fixed_unchecked(row.width) - Row.elementsWidth(row)) /
							2.0,
					);
					break;
				case "Right":
					top_left = top_left.move_x_by(
						Width.get_fixed_unchecked(row.width) - Row.elementsWidth(row),
					);
					break;
				case "Justified":
					per_elem_space =
						(Width.get_fixed_unchecked(row.width) - Row.elementsWidth(row)) /
						(row.elements.length - 1);
					break;
			}

			const renderedElements = [];

			for (const element of row.elements) {
				const result = computeTextboxPositions(element, top_left, font_dict);
				depth = Math.max(depth, result.depth);
				top_left = top_left.move_x_by(
					Width.get_fixed_unchecked(element.width) + per_elem_space,
				);
				renderedElements.push(result.renderedLayout);
			}

			depth += row.margin.bottom;
			return {
				depth,
				renderedLayout: {
					...l,
					bounding_box: new Box(
						originalTopLeft,
						originalTopLeft
							.move_y_by(depth)
							.move_x_by(Width.get_fixed_unchecked(row.width)),
					),
					elements: renderedElements,
				},
			};
		}

		case "Elem": {
			const elem = l as Elem.t;

			if (elem.is_ref) {
				throw new Error(
					"Cannot compute textbox positions of uninstantiated layout",
				);
			}

			const height = Font.get_height(elem.font, font_dict);
			top_left = top_left
				.move_y_by(elem.margin.top)
				.move_x_by(elem.margin.left);
			let line = 1;
			let cursor = top_left.x;
			elem.spans.forEach((span) => {
				if (
					cursor - top_left.x + span.width >
						Width.get_fixed_unchecked(elem.width) - elem.margin.right ||
					span.text === "\n\n"
				) {
					cursor = top_left.x;
					line += 1;
				}
				span.bbox = new Box(
					new Point(cursor - top_left.x, (line - 1) * height),
					new Point(cursor + span.width, line * height),
				);
				span.line = line;
				cursor += span.width;
			});

			const lineWidths = elem.spans.reduce(
				(acc, span) => {
					acc[span.line - 1].width += span.width;
					return acc;
				},
				Array.from({ length: line }, (_, i) => ({ line: i + 1, width: 0 })),
			);

			switch (elem.alignment) {
				case "Center":
					elem.spans.forEach((span) => {
						span.bbox = span.bbox.move_x_by(
							(Width.get_fixed_unchecked(elem.width) -
								lineWidths[span.line - 1].width) /
								2.0,
						);
					});
					break;
				case "Right":
					elem.spans.forEach((span) => {
						span.bbox = span.bbox.move_x_by(
							Width.get_fixed_unchecked(elem.width) -
								lineWidths[span.line - 1].width,
						);
					});
					break;
				case "Justified": {
					for (let i = 1; i < line; i++) {
						const lineSpans = elem.spans.filter(
							(span) =>
								span.line === i &&
								span.text !== "\n\n" &&
								span.text !== "\n" &&
								span.text !== " ",
						);
						const totalWidth = lineSpans.reduce(
							(acc, span) => acc + span.width,
							0,
						);
						const perSpace =
							(Width.get_fixed_unchecked(elem.width) - totalWidth) /
							(lineSpans.length - 1);
						let cursor = 0;
						lineSpans.forEach((span) => {
							span.bbox = span.bbox.move_x_by(cursor - span.bbox.top_left.x);
							cursor += span.width + perSpace;
						});
					}
				}
			}

			const textbox = new Box(
				top_left,
				top_left
					.move_x_by(Width.get_fixed_unchecked(elem.width))
					.move_y_by(height * line + elem.margin.bottom),
			);
			return {
				depth: top_left.y + height * line + elem.margin.bottom,
				renderedLayout: { ...l, bounding_box: textbox },
			};
		}
	}
}

export type Color =
	| "Transparent" // transparent
	| "Light Yellow" // "#FFC96F"
	| "Light Brown" // "#ECB176"
	| "Light Green" // "#F6FAB9"
	| "Light Beige" // "#F6EEC9"
	| "Light Blue" // "#EEF7FF"
	| "Blue"; // "#4793AF"

export const ColorMap = {
	Transparent: "transparent",
	"Light Yellow": "#FFC96F",
	"Light Brown": "#ECB176",
	"Light Green": "#F6FAB9",
	"Light Beige": "#F6EEC9",
	"Light Blue": "#EEF7FF",
	Blue: "#4793AF",
};
